worker/upload.go

```diff
@@ -0,0 +1,1126 @@
+
+			sector:      (*[rhpv2.SectorSize]byte)(shard),
+			sectorIndex: sI,
+		}
+	}
+
+	return slab, uploads, responseChan
+}
+
+func (u *upload) canUseUploader(ul *uploader, sID slabID) bool {
+	u.mu.Lock()
+	defer u.mu.Unlock()
+
+	_, excluded := u.excluded[ul.fcid]
+	if excluded {
+		return false
+	}
+
+	_, used := u.used[sID][ul.fcid]
+	return !used
+}
+
+func (u *upload) start(ctx context.Context, r io.Reader, rs api.RedundancySettings) chan slabResponse {
+	// create the response channel
+	slabsChan := make(chan slabResponse)
+
+	// launch a goroutine uploading the slabs
+	go func() {
+		var wg sync.WaitGroup
+		defer func() {
+			wg.Wait()
+			close(slabsChan)
+		}()
+
+		var slabIndex int
+		for res := range u.read(ctx, r, rs) {
+			if res.err != nil {
+				slabsChan <- slabResponse{err: res.err}
+				return
+			}
+
+			wg.Add(1)
+			go func(data []byte, length, index int) {
+				defer wg.Done()
+
+				// add tracing
+				ctx, span := tracing.Tracer.Start(ctx, "createShards")
+				defer span.End()
+
+				// create slab
+				s := object.NewSlab(uint8(rs.MinShards))
+
+				// create the shards
+				shards := make([][]byte, rs.TotalShards)
+				s.Encode(data, shards)
+				s.Encrypt(shards)
+
+				// upload the shards
+				var err error
+				s.Shards, err = u.uploadShards(ctx, shards, index)
+				if err != nil {
+					select {
+					case slabsChan <- slabResponse{err: err}:
+					default:
+					}
+					return
+				}
+
+				// send the slab
+				select {
+				case slabsChan <- slabResponse{
+					slab: object.SlabSlice{
+						Slab:   s,
+						Offset: 0,
+						Length: uint32(length),
+					},
+					index: index,
+				}:
+				default:
+				}
+			}(res.data, res.length, slabIndex)
+			slabIndex++
+		}
+	}()
+	return slabsChan
+}
+
+func (u *upload) registerUsedQueue(sID slabID, fcid types.FileContractID) {
+	u.mu.Lock()
+	defer u.mu.Unlock()
+
+	_, exists := u.used[sID]
+	if !exists {
+		u.used[sID] = make(map[types.FileContractID]struct{})
+	}
+	u.used[sID][fcid] = struct{}{}
+}
+
+func (u *upload) read(ctx context.Context, r io.Reader, rs api.RedundancySettings) chan slabData {
+	size := int64(rs.MinShards) * rhpv2.SectorSize
+	data := make(chan slabData)
+
+	go func() {
+		for {
+			select {
+			case <-u.mgr.stopChan:
+				data <- slabData{err: errors.New("manager was stopped")}
+				return
+			case <-ctx.Done():
+				data <- slabData{err: errors.New("upload timed out")}
+				return
+			case <-u.nextReadTrigger:
+			}
+
+			buf := make([]byte, size)
+			length, err := io.ReadFull(io.LimitReader(r, size), buf)
+			if err == io.EOF {
+				close(data)
+				return
+			} else if err != nil && err != io.ErrUnexpectedEOF {
+				data <- slabData{err: errors.New("data read failed")}
+				return
+			}
+
+			data <- slabData{data: buf, length: length}
+		}
+	}()
+
+	return data
+}
+
+func (u *upload) triggerDoneShard() {
+	select {
+	case u.doneShardTrigger <- struct{}{}:
+	default:
+	}
+}
+
+func (u *upload) triggerNextRead() {
```

Comment:
I'd double check whether these two helpers really make sense. If you only call them once, it might be worth dropping them.
e.g. `triggerNextRead` could be inlined in `uploadShards` and in `newUpload` you could do a blank send with a comment explaining why you do this right away.

`triggerDoneShard` I think is only called once.